src/aspire/source/image.py

```diff
@@ -736,13 +760,59 @@ class ArrayImageSource(ImageSource):
     if available, is consulted directly by the parent class, bypassing `_images`.
     """
 
-    def __init__(self, im, metadata=None):
+    def __init__(self, im, metadata=None, angles=None):
         """
         Initialize from an `Image` object
         :param im: An `Image` object representing image data served up by this `ImageSource`
         :param metadata: A Dataframe of metadata information corresponding to this ImageSource's images
+        :param angles: Optional n-by-3 array of rotation angles corresponding to `im`.
         """
+
         super().__init__(
             L=im.res, n=im.n_images, dtype=im.dtype, metadata=metadata, memory=None
         )
+
         self._cached_im = im
+
+        # Create filter indices, these are required to pass unharmed through filter eval code
+        #   that is potentially called by other methods later.
+        self.filter_indices = np.zeros(self.n)
+        self.unique_filters = [IdentityFilter()]
+
+        # Optionally populate angles/rotations.
+        if angles is not None:
+            if angles.shape != (self.n, 3):
+                raise ValueError(f"Angles should be shape {(self.n, 3)}")
+            # This will populate `_rotations`,
+            #   which is exposed by properties `angles` and `rots`.
+            self.angles = angles
+
+    def _rots(self):
```

Comment:
Why not put this (and `_angles`) in the `ImageSource` class. Presumably, we'd want this type of behavior whenever rotations are missing.

Reply:
I guess I was trying to balance the scope of the patch.

Yes, it does make sense for `ImageSource` and `ArrayImageSource`. However, I wasn't sure if it always made sense for any subclass. Specifically 2/3 that exist today, `mrcstack` and `RelionSource`.  To accommodate those would require refactoring the classes more.  It looks like it becomes writing a unique `_rots` and `_angles` for each of those remaining subclasses anyway, so it seems there is not much gained today by putting into `ImageSource` at this time.

FWIW, It's sort of older school python, but this is where Mix-Ins can be handy. I think they are frowned upon these days for some reasons... but favored approaches really require we get the details of the abstractions right; this wasn't done and is why I had to add the `_rots` and `_angles` methods... 

I think this is a solid patch covering stuff that was in my way.  Although my dark side thinks we should just eviscerate most of this code, I'm just trying to cover the ArrayImageSource case right now.  I am using this patch already to tinker with Amit H's data.... We can mimic or refactor more to extend further if and when these issues come up again.  Let's talk about it in the meeting tomorrow.

Reply:
I guess what I am saying is I think these source classes are just structured sub optimally. Fixing them all together requires refactoring/rewriting all of them, and is out of scope at this moment.